devtools-html/debugger.html
# [Work in progress] Add reselect and reduce unnecessary renders

Associated Issue: #2727

### Background

In issue #2727, @fflorent reported a very severe performance problem on pages with lots of script tags. In the course of debugging that problem, we noticed that all the components connected to the redux store were being re-rendered on every state change, even if the state change didn't effect the data that component read. That was odd, because Redux promises not to re-render `connect`ed components if the props don't change (as tested by shallow equality). The culprit was our selector functions, which often use `map` and `filter` to build up results. Even if the underlying data stayed the same, these selectors return new objects and arrays every time they are called. A common solution to this problem is to use memoized selectors, which promise to return the same objects if the underlying data hasn't changed. One popular library for doing this is [reselect](https://github.com/reactjs/reselect). 

So the goal of this PR is to make components only re-render when their underlying data has changed, by memoizing selectors as necessary.

### Summary of Changes

For each component, make it a `PureComponent` and ensure its selectors don't change if their underlying data doesn't change:

- [ ] Frames
- [x] Breakpoints
- [x] Scopes
- [x] Expressions

### Test Plan

All existing tests should pass (no functionality should have changed).

To verify that re-renders have decreased:

- Do a development build so React Perf tools are installed.
- Open debugger.html and then open a debugger for it.
- Type `Perf.start()`
- Interact with the app as much as you like...
- Type `Perf.stop()`
- `Perf.printExclusive()` will now print a table with how long it spend re-rendering each component. This allows you to see if one component is changing because of unrelated data.
